xtext-website/documentation/303_runtime_concepts.md

```diff
@@ -632,6 +632,87 @@ public interface ITokenStream {
 }
 ```
 
+## Tree Manipulation {#tree-manipulation}
+An implementation of [IDerivedStateComputer](https://github.com/eclipse/xtext/blob/main/org.eclipse.xtext/src/org/eclipse/xtext/resource/IDerivedStateComputer.java) can be used to modify the objects in the abstract syntax tree after parsing. Commonly, implementations of this interface are used to modify the objects' fields programmatically.
+
+As an example, let's take a simple Entity DSL defined by the following grammar:
+
+```xtext
+grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+
+generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+
+Model:
+    entities+=Entity*;
+
+Entity:
+    'Entity' (name=ID | name=ID? 'overrides' overriddenEntity=[Entity]);
+```
+
+This grammar allows declaring that an Entity overrides another Entity. In this case, by design and for the sake of conciseness it is not required to explicitly name the overriding Entity. However, you might want to reference an unnamed overriding Entity later, e.g. to override it in turn. This is where IDerivedStateComputer comes in handy.
+
+Let's see a sample implementation of this interface that takes care of setting the `name` field of unnamed Entities. We want to set it to the name of the overridden Entity, suffixed with "_overridden". For this purpose, we write the following implementation and place it, for instance, inside the `org.xtext.example.mydsl.services` package of our DSL project:
+
+```java
+package org.xtext.example.mydsl.services;
+
+import org.eclipse.xtext.resource.DerivedStateAwareResource;
+import org.eclipse.xtext.resource.IDerivedStateComputer;
+import org.xtext.example.mydsl.myDsl.Entity;
+
+public class MyDslDerivedStateComputer implements IDerivedStateComputer {
+    private static final String SUFFIX = "_overridden";
+
+    @Override
+    public void installDerivedState(DerivedStateAwareResource resource, boolean preLinkingPhase) {
+        resource.getAllContents().forEachRemaining(eObject -> {
```

Comment:
I don't think this example is especially beginner-friendly. The derived state is installed before the linking and here we rely on linking information to compute a name property, which in itself would be necessary to make the names that are indexed. Can you imagine a different use-case that you'd want to show here?

Reply:
@szarnekow
Thanks for the feedback. I've provided this example because it is very similar to what I am doing with my custom IDerivedStateComputer implementation in the project I'm working on. It seems to "just work" as expected. Are you saying that my example, apart from not being beginner-friendly, is built on wrong assumptions? This would mean that I have to rework both the example and, most importantly, the code in my project.

Reply:
Yes, that's what I fear is the case.

Reply:
Ok. I will think of another example use case for the documentation. Though first I need to figure out how to address this issue inside my project. I already have sketched up an alternative approach that does not rely on indexing/linking-related information, although I will probably need some help to refine it. But this is off-topic here, I'll open a thread in the Discussions section. I hope you won't mind me pinging you there. Thanks a lot for your help!